File: python/sorting/merge sort.py
def merge(arr,beg,mid,end):
    i = beg
    j = mid+1
    temp = []
    while i<=mid and j<=end:
        if arr[i]<arr[j]:
            temp.append(arr[i])
            i += 1
        else:
            temp.append(arr[j])
            j += 1
    while i<=mid:
        temp.append(arr[i])
        i += 1
    while j<=end:
        temp.append(arr[j])
        j += 1
    for i in range(len(temp)):
        arr[beg] = temp[i]
        beg += 1

def mergesort(arr,beg,end):
    if beg<end:
        mid = (beg+end)//2
        mergesort(arr,beg,mid)
        mergesort(arr,mid+1,end)
        merge(arr,beg,mid,end)

arr = list(map(int,input().strip().split()))
mergesort(arr, 0, len(arr)-1)
print(arr)

File: C/sorting/mergesort.c
#include <stdio.h>
#include<conio.h>
void partition(int a[],int low,int high)
{
    int mid;
    if(low < high)
    {
        mid = (low + high) / 2;
        partition(a, low, mid);
        partition(a, mid + 1, high);
        mergeSort(a, low, mid, high);
    }
}
void mergeSort(int a[],int low,int mid,int high)
{
    int i, mi, k, lo, temp[50];
    lo = low;
    i = low;
    mi = mid + 1;
    while ((lo <= mid) && (mi <= high))
    {
        if (a[lo] <= a[mi])
        {
            temp[i] = a[lo];
            lo++;
        }
        else
        {
            temp[i] = a[mi];
            mi++;
        }
        i++;
    }
    if (lo > mid)
    {
        for (k = mi; k <= high; k++)
        {
            temp[i] = a[k];
            i++;
        }
    }
    else
    {
        for (k = lo; k <= mid; k++)
        {
             temp[i] = a[k];
             i++;
        }
    }
    for (k = low; k <= high; k++)
    {
        a[k] = temp[k];
    }
}
int main()
{
    int a[50];
    int i, n;
    printf("Enter number of elements in array :");
    scanf("%d",&n);
    printf("Enter elements of array :\n");
    for(i = 0; i < n; i++)
    {
        printf("a[%d] : ",(i+1));
        scanf("%d",&a[i]);
    }
    partition(a, 0, n - 1);
    printf("\n\nSorted array :  ");
    for(i = 0;i < n; i++)
    {
         printf("%d   ",a[i]);
    }
    getch();
    return 0;
}
